### ISSUE TITLE
resource_update does not handle free-form resource extras

### ISSUE THREAD

**Issue by amercader** [ckan/ckan#2290]
**TL;DR**: `resource_update` ignores custom fields in resource. If you update both core and custom fields there is a resource revision created with all the correct changes but not only the core fields are updated on the main resource table. Also updating just the custom field is completely ignored (ie no revision created).

This is blocking #2234 (with effects on #2204)

-----

Let's create a dataset

```
ckanapi action package_create name=test-resource-extras
```

Now create a resource for it with a custom field:
```
ckanapi action resource_create package_id=f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0 name=test-resource format=txt url=url1 custom_field=1234

{
  "custom_field": "1234",
  "format": "TXT",
  "id": "29d608f1-7720-4430-b2db-8d221ab76974",
  "package_id": "f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0",
  "url": "http://url1",
...
}

```
Cool, the response contains the custom field.

Just to confirm, calling package_show (which goes through the search index) also shows de custom field:

```
ckanapi action package_show id=f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0

{
...
  "id": "f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0",
  "resources": [
    {
      "custom_field": "1234",
      "format": "TXT",
      "hash": "",
      "id": "29d608f1-7720-4430-b2db-8d221ab76974",
      "package_id": "f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0",
      ...
}
```

At this point the database shows:

```
ckan_default=# select package_id, name, extras, url from resource where package_id ='f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0';
              package_id              |     name      |          extras          | url  
--------------------------------------+---------------+--------------------------+------
 f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0 | test-resource | {"custom_field": "1234"} | url1
(1 row)

ckan_default=# select package_id, name, extras, url from resource_revision where package_id ='f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0';
              package_id              |     name      |          extras          | url  
--------------------------------------+---------------+--------------------------+------
 f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0 | test-resource | {"custom_field": "1234"} | url1
(1 row)

```

So far so good.

## 1st problem
Let's update the resource (core fields and custom field):
```
ckanapi action resource_update id=29d608f1-7720-4430-b2db-8d221ab76974 package_id=0f1b0197-f4ae-4db1-abd5-065e7147f7a0 name=test-resource-updated format=txt url=url2 custom_field=5678

{
  "custom_field": "1234",
  "format": "TXT",
  "id": "29d608f1-7720-4430-b2db-8d221ab76974",
  "name": "test-resource-updated",
  "url": "http://url2",
  ...
}
```

Core fields were updated, but not the custom field. 

The db shows the following:

```
ckan_default=# select package_id, name, extras, url from resource where package_id ='f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0';
              package_id              |         name          |          extras          | url  
--------------------------------------+-----------------------+--------------------------+------
 f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0 | test-resource-updated | {"custom_field": "1234"} | url2
(1 row)

ckan_default=# select package_id, name, extras, url, revision_timestamp from resource_revision where package_id ='f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0';
              package_id              |         name          |          extras          | url  |     revision_timestamp     
--------------------------------------+-----------------------+--------------------------+------+----------------------------
 f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0 | test-resource         | {"custom_field": "1234"} | url1 | 2015-02-16 15:12:20.229686
 f01e691e-eb7b-40c7-9a12-a95dc8e4a7c0 | test-resource-updated | {"custom_field": "5678"} | url2 | 2015-02-16 15:15:52.859463

```
So there is a more recent revision that contains the updated extras with the custom field, but this has not been applied to the resource table.

## 2nd problem

We update the same resource, but just updating the custom field:
```
ckanapi action resource_update id=29d608f1-7720-4430-b2db-8d221ab76974 package_id=0f1b0197-f4ae-4db1-abd5-065e7147f7a0 name=test-resource-updated format=txt url=url2 custom_field=aaaa

{
  "custom_field": "1234",
  "format": "TXT",
  "id": "29d608f1-7720-4430-b2db-8d221ab76974",
  "name": "test-resource-updated",
  "url": "http://url2",
  ...
}
```

There is a successful response, but the custom field was not updated, and the database does not show a new revision for the change:

```
ckan_default=# select package_id, name, extras, url, revision_timestamp from resource_revision where package_id ='0f1b0197-f4ae-4db1-abd5-065e7147f7a0';
              package_id              |         name          |          extras          | url  |     revision_timestamp     
--------------------------------------+-----------------------+--------------------------+------+----------------------------
 0f1b0197-f4ae-4db1-abd5-065e7147f7a0 | test-resource         | {"custom_field": "1234"} | url1 | 2015-02-16 15:12:20.229686
 0f1b0197-f4ae-4db1-abd5-065e7147f7a0 | test-resource-updated | {"custom_field": "5678"} | url2 | 2015-02-16 15:15:52.859463
(2 rows)

```

**Comment by wardi** [comment]
@amercader dupe of #2158 but if you want it you can have it :-)

**Comment by amercader** [comment]
argh, I missed that one, it would have save a couple of hours of debugging :(
I'll have another look with the pointers on that PR.

Closing in favour of #2158
